apps/solaredgemonitor/solaredge_monitor.star

```diff
@@ -5,167 +5,671 @@ Description: Energy production and consumption monitor for your SolarEdge solar
+""")
+
+def render_fail(rep):
+    content = json.decode(rep.body())
+    return render.Root(render.Box(render.WrappedText(str(rep.status_code) + " : " + content["String"], color = RED)))
+
+def get_energy_details(site_id, api_key, tz, interval):
+    now = time.now().in_location(tz)
+    start_string = ""
+    now_string = humanize.time_format("yyyy-MM-dd HH:mm:ss", now)
+    if interval == "day":  # from the start of today to now, just set time to 00:00
+        start_string = humanize.time_format("yyyy-MM-dd 00:00:00", now)
+        now_string = humanize.time_format("2006-01-02 15:04:05", now)
+        time_unit = "DAY"
+    elif interval == "24h":  # the last 24hr from now, use a now-duration
+        duration = time.parse_duration("24h")
+        start_time = now - duration
+        start_string = humanize.time_format("yyyy-MM-dd HH:mm:ss", start_time)
+        time_unit = "DAY"
+    elif interval == "month":  # from the start of the month to now, just set day to 01 and time to 00:00
+        month = now.month
+        if now.month < 10:  # pad a zero if less then 10
+            month = "0" + str(month)
+        start_string = "{}-{}-01 00:00:00".format(now.year, month)
+        time_unit = "MONTH"
+    elif interval == "year":  # from the start of the year to now, set month/day to 01/01 and time to 00:00
+        start_string = "{}-01-01 00:00:00".format(now.year)
+        time_unit = "YEAR"
+    else:
+        return None
+    url = URL_AUT.format(site_id, start_string, now_string)
+    rep = http.get(
+        url,
+        params = {
+            "api_key": api_key,
+            "startTime": start_string,
+            "endTime": now_string,
+            "timeUnit": time_unit,
+        },
+        ttl_seconds = CACHE_TTL,
+    )
+
+    if rep.status_code != 200:
+        print(rep.status_code)
+        print(rep.body())
+        return None
+
+    return rep.json()["energyDetails"]
+
+def get_autarky_percent(site_id, api_key, tz, interval):
+    energy_details = get_energy_details(site_id, api_key, tz, interval)
+    if not energy_details:
+        return 0
+    self_consumption = 0
+    consumption = 0
+    for m in energy_details["meters"]:
+        if m["type"] == "SelfConsumption":
+            for v in m["values"]:
+                self_consumption += v["value"]
+        elif m["type"] == "Consumption":
+            for v in m["values"]:
+                consumption += v["value"]
+    if consumption == 0:
+        return 100
+    return int(self_consumption / consumption * 100)
 
 def main(config):
     api_key = config.str("api_key")
     site_id = humanize.url_encode(config.str("site_id", ""))
+    tz = config.get("$tz", "Europe/Zurich")
```

Comment:
Seems like a little arbitrary as a default, how about defaulting to the system's time zone? Or `UTC`, if there isn't a way to get the system's time zone.

Reply:
Good idea. I've updated the PR to retrieve the time zone from the site details API and fall back to UTC if it's not available. However, it looks like the time zone is not part of the v2 schema (https://developers.solaredge.com/docs/monitoring/ed297a6bcaeb6-site-details), so that part will likely go away again, when this app is ported to the v2 API.